File: angular/projects/site/src/lib/field-types/inline-field-definition.ts
/**
 * One field defined inline as part of a composite field type's own configuration - a `Matrix` block
 * type's sub-field, or a `Table` column. Mirrors `InlineFieldDefinition`
 * (`src/Dignite.FlexFields.Site/Dignite/FlexFields/Site/InlineFieldDefinition.cs`), shared between the
 * two rather than each declaring its own copy.
 *
 * Camel-cased throughout, matching how the server's `JsonSerializerDefaults.Web` always writes it back
 * out - nothing here is ever authored in a different casing first, so there is no dual-casing round trip
 * to normalize (unlike `Select.Options`/`SelectListItem`).
 */
export interface InlineFieldDefinition {
  name: string;
  displayName: string;
  description?: string;
  fieldTypeName: string;
  required: boolean;
  configuration: Record<string, unknown>;
}

function normalizeInlineFieldDefinition(item: unknown): InlineFieldDefinition {
  const source = (item ?? {}) as Partial<InlineFieldDefinition>;
  return {
    name: source.name ?? '',
    displayName: source.displayName ?? '',
    description: source.description,
    fieldTypeName: source.fieldTypeName ?? '',
    required: source.required ?? false,
    configuration: (source.configuration as Record<string, unknown>) ?? {},
  };
}

/** Reads a stored `Matrix.BlockTypes[].fields` or `Table.Columns` configuration value, defensively. */
export function normalizeInlineFieldDefinitions(source: unknown): InlineFieldDefinition[] {
  return Array.isArray(source) ? source.map(normalizeInlineFieldDefinition) : [];
}

File: src/Dignite.FlexFields.Site/Dignite/FlexFields/Site/InlineFieldDefinition.cs
using Dignite.Abp.FlexFields;

namespace Dignite.FlexFields.Site;

/// <summary>
/// One field defined inline as part of a composite field type's own configuration - a <c>Matrix</c>
/// block type's sub-field, or a <c>Table</c>'s column. There is no persisted identity of its own
/// (contrast <see cref="IFlexFieldData"/>, a top-level field's definition, which does), and it is
/// shared verbatim between the two rather than each declaring its own copy.
///
/// <para>
/// Deliberately not just a <see cref="FlexFieldData"/>: that type carries no <c>Required</c> flag,
/// because in the kernel proper "required" is a property of a field's <i>usage</i>
/// (<see cref="FlexFieldValue.Required"/>), not its definition - a definition can be attached to
/// several host types with different Required settings. An inline field has no separate usage record
/// to carry that flag instead, so it lives directly here.
/// </para>
/// </summary>
public class InlineFieldDefinition
{
    public string Name { get; set; } = default!;

    public string DisplayName { get; set; } = default!;

    public string? Description { get; set; }

    /// <summary>Name of the <see cref="IFieldType"/> this field is bound to.</summary>
    public string FieldTypeName { get; set; } = default!;

    public bool Required { get; set; }

    public FieldConfigurationDictionary Configuration { get; set; } = new();
}
